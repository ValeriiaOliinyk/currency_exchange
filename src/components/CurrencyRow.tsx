import React, { useState, useEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import PropTypes from "prop-types";
import { Formik } from "formik";
import {
  loadData,
  getCurrencyArray,
  getFromCurrency,
  getToCurrency,
  getExchangeRate,
  updateData,
  getUpdatedData,
  addExchangeRate,
} from "../redux/ducks/currency";

import { validationSchema } from "../helpers/validation";
import { InitialValues, CurrencyTypes } from "../helpers/interfaces";

// Components
import {
  FormBox,
  Option,
  FormEqually,
  FormCurrency,
  FormControl,
  FormSelect,
  ErrorMessage,
} from "../styles/components";

// Styles
import { Container } from "react-bootstrap";

export function CurrencyRow() {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(loadData());
  }, [dispatch]);

  const currency: Array<string> | undefined = useSelector(getCurrencyArray);
  const selectedFromCurrency: string | undefined = useSelector(getFromCurrency);
  const selectedToCurrency: string | undefined = useSelector(getToCurrency);
  const selectedExchangeRate: number | undefined = useSelector(getExchangeRate);
  const updatedData: CurrencyTypes = useSelector(getUpdatedData);

  const [fromCurrency, setFromCurrency] = useState<string | undefined>(
    selectedFromCurrency
  );
  const [toCurrency, setToCurrency] = useState<string | undefined>(
    selectedToCurrency
  );
  const [exchangeRate, setExchangeRate] = useState<number>(
    selectedExchangeRate
  );
  const [amount, setAmount] = useState<number>(1);
  const [amountInFromCurrency, setAmountInFromCurrency] = useState<boolean>(
    true
  );

  const firstExchange = useRef<number>(exchangeRate);

  useEffect(() => {
    setToCurrency(selectedToCurrency);
  }, [selectedToCurrency]);

  useEffect(() => {
    if (updatedData.rates && toCurrency) {
      setExchangeRate(updatedData.rates[toCurrency]);
    }
  }, [dispatch, toCurrency, updatedData.rates]);

  useEffect(() => {
    dispatch(addExchangeRate(firstExchange.current));
  }, [dispatch]);

  useEffect(() => {
    if (fromCurrency && toCurrency) {
      dispatch(updateData(fromCurrency, toCurrency));
    }
  }, [dispatch, fromCurrency, toCurrency]);

  const getInitialValues = (): InitialValues => {
    return {
      fromAmount: amountInFromCurrency
        ? amount
        : (amount / exchangeRate).toFixed(1),
      toAmount: amountInFromCurrency
        ? (amount * exchangeRate).toFixed(1)
        : amount,
    };
  };

  return (
    <Container>
      <Formik
        enableReinitialize={true}
        initialValues={getInitialValues()}
        validationSchema={validationSchema}
        onSubmit={() => {}}
      >
        {({ values, errors, touched }) => (
          <FormCurrency>
            <FormBox>
              <FormControl
                type="number"
                name="fromAmount"
                value={values.fromAmount}
                onChange={(e: {
                  target: { value: React.SetStateAction<number> };
                }) => {
                  setAmount(e.target.value);
                  setAmountInFromCurrency(true);
                }}
              />
              <FormSelect
                as="select"
                value={fromCurrency}
                onChange={(e: {
                  target: { value: React.SetStateAction<string | undefined> };
                }) => setFromCurrency(e.target.value)}
              >
                {currency &&
                  currency.map((item) => (
                    <Option key={item} value={item}>
                      {item}
                    </Option>
                  ))}
              </FormSelect>
            </FormBox>
            {errors.fromAmount && touched.fromAmount ? (
              <ErrorMessage>{errors.fromAmount}</ErrorMessage>
            ) : (
              <ErrorMessage transparent>Transparent</ErrorMessage>
            )}
            <FormEqually />
            <FormBox>
              <FormControl
                type="number"
                name="toAmount"
                value={values.toAmount}
                onChange={(e: {
                  target: { value: React.SetStateAction<number> };
                }) => {
                  setAmount(e.target.value);
                  setAmountInFromCurrency(false);
                }}
              />
              <FormSelect
                as="select"
                value={toCurrency}
                onChange={(e: {
                  target: { value: React.SetStateAction<string | undefined> };
                }) => setToCurrency(e.target.value)}
              >
                {currency &&
                  currency.map((item) => (
                    <Option key={item} value={item}>
                      {item}
                    </Option>
                  ))}
              </FormSelect>
            </FormBox>
            {errors.toAmount && touched.toAmount ? (
              <ErrorMessage>{errors.toAmount}</ErrorMessage>
            ) : (
              <ErrorMessage transparent>Transparent</ErrorMessage>
            )}
          </FormCurrency>
        )}
      </Formik>
    </Container>
  );
}

CurrencyRow.defaultProps = {
  currency: [],
  selectedFromCurrency: "",
  selectedToCurrency: "",
  selectedExchangeRate: 0,
};

CurrencyRow.propTypes = {
  currency: PropTypes.arrayOf(PropTypes.string),
  selectedFromCurrency: PropTypes.string,
  selectedToCurrency: PropTypes.string,
  selectedExchangeRate: PropTypes.number,
};
