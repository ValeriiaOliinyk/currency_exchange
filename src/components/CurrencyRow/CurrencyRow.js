import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  getCurrencyArray,
  getFromCurrency,
  getToCurrency,
  getExchangeRate,
} from "../../redux/currency/currency-selectors";
import { loadData, updateData } from "../../redux/currency/currency-reducers";
import axios from "axios";

const BASE_URL = "https://api.exchangeratesapi.io/latest";

export default function CurrencyRow() {
  const dispatch = useDispatch();

  const currency = useSelector(getCurrencyArray);
  const selectedFromCurrency = useSelector(getFromCurrency);
  const selectedToCurrency = useSelector(getToCurrency);
  const selectedExchangeRate = useSelector(getExchangeRate);

  const [fromCurrency, setFromCurrency] = useState("");
  const [toCurrency, setToCurrency] = useState("");
  const [exchangeRate, setExchangeRate] = useState(0);
  const [amount, setAmount] = useState(1);
  const [amountInFromCurrency, setAmountInFromCurrency] = useState(true);

  useEffect(() => {
    dispatch(loadData());
  }, [dispatch]);

  useEffect(() => {
    if (selectedFromCurrency && selectedToCurrency) {
      setFromCurrency(selectedFromCurrency);
      setToCurrency(selectedToCurrency[0]);
      setExchangeRate(selectedExchangeRate);
    }
  }, [selectedExchangeRate, selectedFromCurrency, selectedToCurrency]);

  let fromAmount = null;
  let toAmount = null;

  if (amountInFromCurrency) {
    fromAmount = amount;
    toAmount = (amount * exchangeRate).toFixed(1);
  } else {
    toAmount = amount;
    fromAmount = (amount / exchangeRate).toFixed(1);
  }

  const handleFromAmountChange = (e) => {
    setAmount(e.target.value);
    setAmountInFromCurrency(true);
  };

  const handleToAmountChange = (e) => {
    setAmount(e.target.value);
    setAmountInFromCurrency(false);
  };

  useEffect(() => {
    if (fromCurrency && toCurrency) {
      axios(`${BASE_URL}?base=${fromCurrency}&symbols=${toCurrency}`)
        .then(({ data }) => setExchangeRate(data.rates[toCurrency]))
        .catch((error) => console.log(error));
    }
    dispatch(updateData(fromCurrency, toCurrency));
  }, [dispatch, fromCurrency, toCurrency]);

  return (
    <div>
      <input
        type="number"
        value={fromAmount}
        onChange={handleFromAmountChange}
      />
      <select
        value={fromCurrency}
        onChange={(e) => setFromCurrency(e.target.value)}
      >
        {currency &&
          currency.map((item) => (
            <option key={item} value={item}>
              {item}
            </option>
          ))}
      </select>
      <div>=</div>
      <input type="number" value={toAmount} onChange={handleToAmountChange} />
      <select
        value={toCurrency}
        onChange={(e) => setToCurrency(e.target.value)}
      >
        {currency &&
          currency.map((item) => (
            <option key={item} value={item}>
              {item}
            </option>
          ))}
      </select>
    </div>
  );
}
