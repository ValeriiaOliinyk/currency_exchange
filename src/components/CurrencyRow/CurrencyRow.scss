.Form {
  margin: auto;
  width: 500px;
  height: 180px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.75);
  padding: 20px;
  border-radius: 20px;
}

.Form__box {
  display: flex;
  width: 100%;
  justify-content: space-between;
}

.Form__equally {
  text-align: center;
  font-size: 30px;
}

.Form__input {
  width: 75%;
}

.Form__select {
  width: 20%;
}
