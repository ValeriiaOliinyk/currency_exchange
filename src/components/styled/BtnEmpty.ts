import styled from "styled-components";

interface BtnEmptyProps {
  type: string;
}

export const BtnEmpty = styled.section<BtnEmptyProps>`
  @media ${({ theme }) => theme.breakpoints.tablet} {
    width: 15px;
    height: 15px;
  }
  width: 20px;
  height: 20px;
  background-image: url(${`${window.location.origin}/images/starEmpty.svg`});
  background-repeat: no-repeat;
  background-size: cover;
`;
