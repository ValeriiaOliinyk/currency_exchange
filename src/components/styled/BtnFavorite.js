import styled from "styled-components";
import star from "../../images/star.svg";

export default styled.section`
  width: 20px;
  height: 20px;
  background-image: url(${star});
  background-repeat: no-repeat;
  background-size: cover;
`;
