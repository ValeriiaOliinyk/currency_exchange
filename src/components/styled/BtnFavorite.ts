import styled from "styled-components";

interface BtnFavoriteProps {
  type: string;
}

export const BtnFavorite = styled.section<BtnFavoriteProps>`
  width: 20px;
  height: 20px;
  background-image: url(${`${window.location.origin}/images/star.svg`});
  background-repeat: no-repeat;
  background-size: cover;
`;
