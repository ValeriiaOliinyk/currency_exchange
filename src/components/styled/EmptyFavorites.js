import styled from "styled-components";

export default styled.li`
  text-align: center;
  font-size: 20px;
  color: ${(props) => props.theme.colors.text};
  font-weight: 800;
`;
