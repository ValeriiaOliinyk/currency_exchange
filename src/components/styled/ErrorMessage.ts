import styled from "styled-components";

interface StyledProps {
  transparent?: boolean;
}

export const ErrorMessage = styled.div<StyledProps>`
  width: 100%;
  font-size: 18px;
  font-weight: 700;
  color: ${(props) =>
    props.transparent
      ? props.theme.colors.transparent
      : props.theme.colors.red};
`;
