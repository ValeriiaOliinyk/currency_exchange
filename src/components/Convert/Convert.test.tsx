import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import { Container, Title } from "../../styled";
import { ThemeProvider } from "styled-components";
import Theme from "../../styles/theme";
import { Provider } from "react-redux";
import store from "../../redux/store";

import { CurrencyRow } from "../CurrencyRow/CurrencyRow";

import theme from "../../styles/theme";
import { shallow } from "enzyme";

describe("Convert component renders correctly", () => {
  it("CurrencyRow component works", () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={Theme}>
          <Provider store={store.store}>
            <CurrencyRow />
          </Provider>
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Container component works", () => {
    const tree = shallow(<Container />);
    expect(tree).toMatchSnapshot();
  });

  it("Renders correctly", () => {
    const tree = renderer.create(<Title theme={theme}>Convert</Title>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
