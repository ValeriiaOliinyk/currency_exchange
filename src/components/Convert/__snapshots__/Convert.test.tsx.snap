// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert component renders correctly Container component works 1`] = `
.c0 {
  width: 80%;
  margin: auto;
  margin-top: 100px;
}

@media screen and (max-width:768px) {

}

@media screen and (max-width:425px) {

}

@media screen and (max-width:768px) {

}

<section
  className="c0"
/>
`;

exports[`Convert component renders correctly CurrencyRow component works 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  outline: none;
  width: 75%;
  padding: 10px;
  border-radius: 6px;
  border: 1px solid #000000;
  border-color: #000000;
}

.c0 {
  margin: auto;
  width: 50%;
  height: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
  padding: 20px;
  border-radius: 20px;
}

.c5 {
  width: 40px;
  height: 40px;
  background-image: url(equal.svg);
  text-align: center;
  font-size: 30px;
  margin-bottom: 26px;
}

.c3 {
  width: 20%;
}

.c4 {
  width: 100%;
  font-size: 18px;
  font-weight: 700;
  color: #ffffff00;
}

@media screen and (max-width:768px) {
  .c0 {
    width: 90%;
  }
}

@media screen and (max-width:425px) {
  .c3 {
    width: 40%;
  }
}

@media screen and (max-width:768px) {
  .c3 {
    font-size: 12px;
  }
}

<div
  className="container"
>
  <form
    className="c0"
  >
    <div
      className="c1"
    >
      <input
        className="c2"
        name="fromAmount"
        onBlur={[Function]}
        onChange={[Function]}
        type="number"
        value={1}
      />
      <select
        className="c3"
        onChange={[Function]}
      />
    </div>
    <div
      className="c4"
    >
      Transparent
    </div>
    <div
      className="c5"
    />
    <div
      className="c1"
    >
      <input
        className="c2"
        name="toAmount"
        onBlur={[Function]}
        onChange={[Function]}
        type="number"
        value="0.0"
      />
      <select
        className="c3"
        onChange={[Function]}
      />
    </div>
    <div
      className="c4"
    >
      Transparent
    </div>
  </form>
</div>
`;

exports[`Convert component renders correctly renders correctly 1`] = `
.c0 {
  text-align: center;
  margin-bottom: 30px;
  color: #67a030;
  font-weight: 700;
  font-size: 32px;
}

@media screen and (max-width:768px) {
  .c0 {
    font-size: 22px;
  }
}

<h2
  className="c0"
>
  Convert
</h2>
`;
